### PR TITLE
Update Calico confd base image from bacongobbler v0.12.1

### PR DESCRIPTION
**Pull request by robbrockbank**
This PR updates our confd to be based from bacongobbler v0.12.1.  There are three commits from the current v0.12.1 for:
-  Merging k8s backend into confd
-  Adding glide based vendoring and containerized builds
-  Fixing an issue with the etcd backend (which I intend to upstream)

Note that this PR is against a v0.12.1-calico branch (which is currently just the v0.12.1 bacongobbler branch).  The way I've done this I've lost history for the k8s stuff, so i may need to handle this all differently.  WDYT?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# owned by the current user.
+MY_UID:=$(shell id -u)
+MY_GID:=$(shell id -g)
+
+DOCKER_GO_BUILD := mkdir -p .go-pkg-cache && \
+                   docker run --rm \
+                              --net=host \
+                              $(EXTRA_DOCKER_ARGS) \
+                              -e LOCAL_USER_ID=$(MY_UID) \
+                              -v $${PWD}:/go/src/github.com/kelseyhightower/confd:rw \
```
Comment (tomdee) [inline Makefile]: Use CURDIR instead of PWD

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,60 @@
+
+# Disable make's implicit rules, which are not useful for golang, and slow down the build
+# considerably.
+.SUFFIXES:
+
+all: bin/confd bin/confd.static
```
Comment (tomdee) [inline Makefile]: Why have static and dynamically linked versions?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+GO_BUILD_CONTAINER?=calico/go-build:v0.4
+
+# All go files.
+GO_FILES:=$(shell find . -type f -name '*.go') $(GENERATED_GO_FILES)
+
+# Figure out the users UID/GID.  These are needed to run docker containers
+# as the current user and ensure that files built inside containers are
+# owned by the current user.
+MY_UID:=$(shell id -u)
+MY_GID:=$(shell id -g)
```
Comment (tomdee) [inline Makefile]: Is this actually used anywhere

File: Makefile (hunk 2 above)
Comment (robbrockbank) [inline Makefile]: That's a good question.  No idea.  I meant to follow up to see if you knew why we had both.

I'd prefer to just have the static build and get rid of the .static suffix.

File: Makefile (hunk 3 above)
Comment (robbrockbank) [inline Makefile]: nope.  I shall remove.

File: Makefile (hunk 1 above)
Comment (robbrockbank) [inline Makefile]: ta.  will update.

**Comment by heschlie** [comment]
LGTM

**Pull request merged**